### ISSUE TITLE
[Safari] The first line of each paragraph cannot be input into Chinese

### ISSUE THREAD

**Issue by gongzhiq** [ckeditor/ckeditor5#1333]
## Is this a bug report or feature request? (choose one)

🐞 Bug report  

> The first line of each paragraph cannot be input into Chinese.

## 💻 Version of CKEditor

ckeditor5 v11.1.1 classic editor build (latest version)

## 📋 Steps to reproduce

![small6](https://user-images.githubusercontent.com/5800062/47849616-5530c200-de0d-11e8-8712-f597298db5cb.gif)

use Safari Browser
Edition 12.0.1 (13606.2.104.1.2)

**Comment by gongzhiq** [comment]
Hello, when can I repair it? @Reinmar

**Comment by gongzhiq** [comment]
Hello! Yes, there will be a problem at the beginning of each line in simplified Chinese.

![small7](https://user-images.githubusercontent.com/5800062/48056278-63575780-e1ec-11e8-8b1e-9f32e50748f8.gif)

**Comment by gongzhiq** [comment]
Hello, @Reinmar ,I know how to solve this problem, but I don't know how to modify the code.

Listening for input events in text input boxes triggers input events when spelling Chinese characters (input method) but Chinese characters are not actually filled into text boxes (selected words), as shown in Figure 1:

![1069543-20171009211759324-6471773](https://user-images.githubusercontent.com/5800062/48107249-d1962b80-e279-11e8-8b1a-ac284310383f.gif)

Terms of settlement：After the completion of the selection, the input event is triggered, triggering only once.

![1069543-20171009211830262-1175113284](https://user-images.githubusercontent.com/5800062/48107309-1de16b80-e27a-11e8-97f0-c41c3bc3efb5.gif)

```
editor.editing.view.document.on( 'compositionstart', function() {
	$IsInputZh = true;
} );
editor.editing.view.document.on( 'compositionend', function() {
	$IsInputZh = false;
} );
editor.editing.view.document.on( 'input', function() {
       if ($IsInputZh){
                   return;
         }

     // Input content，
} );
```

**Comment by xinglie** [comment]
@gongzhiq I think this is a ckeditor bug , whatever use which input method . I guess ckeditor process content immediately when input event triggered. 
ckeditor may process input event like native input element

**Comment by gongzhiq** [comment]
@xinglie Yes, I think so too. It's a problem when the first paragraph of the text starts typing. It doesn't wait for the Pingyin input to complete, and then it processes the input.

**Comment by Reinmar** [comment]
I'm trying to understand one thing first – the bug was reported for Safari only. I can't reproduce any issue on other browsers (works fine in Chrome and Firefox). Nor, with other languages (checked Hiragana and Korean. Do you confirm that or are we talking about different issues?

Regarding debugging:

1. Typing is handled by this package: https://github.com/ckeditor/ckeditor5-typing
2. Specifically, by the `Input` plugin: https://github.com/ckeditor/ckeditor5-typing/blob/master/src/input.js
3. And even more specifically by this observer: https://github.com/ckeditor/ckeditor5-typing/blob/master/src/utils/injecttypingmutationshandling.js

How it works is:

1. It listens to mutations: https://github.com/ckeditor/ckeditor5-typing/blob/c659fa9368ad229de6fc8c4e78a6ef7b6afb4c46/src/utils/injecttypingmutationshandling.js#L20-L24
2. When a mutation occurs it checks what changed: https://github.com/ckeditor/ckeditor5-typing/blob/c659fa9368ad229de6fc8c4e78a6ef7b6afb4c46/src/utils/injecttypingmutationshandling.js#L62-L72
3. If it's a text insertion or replacement, the `input` command is executed with this data. It's executed in multiple places in that listener, but here's one example: https://github.com/ckeditor/ckeditor5-typing/blob/c659fa9368ad229de6fc8c4e78a6ef7b6afb4c46/src/utils/injecttypingmutationshandling.js#L218-L222
4. The input command changes the model. The model is converted to the view (virtual DOM). If that view differs from the DOM (e.g. the character was inserted by the browser in an incorrect place), we re-render the DOM. We do not touch it otherwise.

Unless... we're talking about composition. Here it's far more tricky. We can't be doing step 4. during composition because touching the DOM means breaking the composition. That's most likely what happens in this case.

So what we do? We listen to composition events and don't do unsafe changes when the composition  is taking place. But currently, that's just in one place: https://github.com/ckeditor/ckeditor5-typing/blob/master/src/utils/injectunsafekeystrokeshandling.js#L45

Most likely, there's some other change in the model/view which should not happen in the case that's broken for Chinese. My guess would be – the `<br>` filler (which is present in empty block elements ) gets removed when you type the first letter and that may break the composition. 

The `<br>` filler rendering is handled by the renderer (https://github.com/ckeditor/ckeditor5-engine/blob/master/src/view/renderer.js) based on what the DOM converter returns (https://github.com/ckeditor/ckeditor5-engine/blob/master/src/view/domconverter.js).

Ticket regarding blocking block filler removal during composition: https://github.com/ckeditor/ckeditor5-engine/issues/1353

It'd be first good to make sure that it's the removal of that `<br>` what breaks the composition. The we could think how to prevent its removal.

**Comment by gongzhiq** [comment]
hi, @Reinmar ,Thank you very much for your answer. Yes, this question only exists in Safari.The above answer is very reasonable. Can we repair it according to the above statement?

**Comment by Reinmar** [comment]
> Can we repair it according to the above statement?

Sure. But what I wrote is just the basics of this entire system, so I didn't even mention how the solution could look like. I just don't know without digging deep into this. 

Anyway, the first thing to do is understanding at what stage the composition gets broken. To do that you need to debug step by step what happens in the renderer after keys are pressed.

**Comment by 275208060** [comment]
![video_2019-07-25_190749000000-000019](https://user-images.githubusercontent.com/1212176/61870419-a8552280-af10-11e9-925a-119e187ea777.gif)
Enter Chinese Link Text also has this bug

**Comment by farthinker** [comment]
@275208060 It has been confirmed as another issue: https://github.com/ckeditor/ckeditor5-typing/issues/149

**Comment by farthinker** [comment]
> ![video_2019-07-25_190749000000-000019](https://user-images.githubusercontent.com/1212176/61870419-a8552280-af10-11e9-925a-119e187ea777.gif)
> Enter Chinese Link Text also has this bug

@275208060 It has been confirmed as another issue: https://github.com/ckeditor/ckeditor5-typing/issues/149

**Comment by Reinmar** [comment]
I'm afraid that since we had to revert the patch, this issue is still present. More details here: https://github.com/ckeditor/ckeditor5/issues/4533#issuecomment-544500122.

**Comment by LukaszGudel** [comment]
Current state of this issue:

*   when you type in new line or in empty table cell then first character is doubled and it's not part of the composition, so it stays after text convertion
*   same behaviour with Pinyin or Hiragana
*   happens only on Safari

![](https://user-images.githubusercontent.com/58658048/104173195-b4cf8280-5405-11eb-9e8c-c743ff9501dd.gif)

**Comment by mlewand** [comment]
#8759 was marked as a duplicate of this issue.

**Comment by yangguansen** [comment]
Does this problem had been solved?

**Comment by yangguansen** [comment]
> Hello, @Reinmar ,I know how to solve this problem, but I don't know how to modify the code.
> 
> Listening for input events in text input boxes triggers input events when spelling Chinese characters (input method) but Chinese characters are not actually filled into text boxes (selected words), as shown in Figure 1:
> 
> ![1069543-20171009211759324-6471773](https://user-images.githubusercontent.com/5800062/48107249-d1962b80-e279-11e8-8b1a-ac284310383f.gif)
> 
> Terms of settlement：After the completion of the selection, the input event is triggered, triggering only once.
> 
> ![1069543-20171009211830262-1175113284](https://user-images.githubusercontent.com/5800062/48107309-1de16b80-e27a-11e8-97f0-c41c3bc3efb5.gif)
> 
> ```
> var $IsInputZh = false;
> 
> editor.editing.view.document.on( 'compositionstart', function() {
> 	$IsInputZh = true;
> } );
> editor.editing.view.document.on( 'compositionend', function() {
> 	$IsInputZh = false;
> } );
> editor.editing.view.document.on( 'input', function() {
>        if ($IsInputZh){
>                return;
>         }
> 
>      // Input content，
> } );
> ```

大哥，这个问题你解决了吗？我看官网demo上没有这个问题了

**Comment by Reinmar** [comment]
@ckeditor/qa-team, could you retest this issue?

**Comment by Mgsy** [comment]
I'm closing it., as it is no longer reproducible.
